### ISSUE TITLE
[sklearn#34071] Opt-in to using thread pools for all parallelism (or almost all)

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#357]
**Mirrored from [scikit-learn#34071](https://github.com/scikit-learn/scikit-learn/issues/34071)**

Original title: Opt-in to using thread pools for all parallelism (or almost all)
Original labels: New Feature, Needs Decision, free-threading
Created: 2026-05-20T20:20:11Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the workflow you want to enable

**User perspective:** By default, most parallelism in sklearn uses process pools. On free-threaded Python, however, threading may well have lower overhead: certainly less memory, but quite possibly it will be faster as well.

> There are exceptions. E.g. as discussed in #34070, parallel algorithms that can only use a limited number of cores may be better run as process pools.

The user experience should ideally be "use a thread pool if it's known to be faster." But it's difficult to be sure in the short term whether that's true without further work. And maybe it's situational.

Until sklearn reaches the point where it's automatically choosing the fastest option, in the short term a user would want to be able _try_ as easily as possible to see if threads everywhere speed things up.

**Maintainer perspective:** When testing for thread-safety, or benchmarking for speed, it's useful to be able to have all parallel algorithms use thread pools by default, whether in a CI job or in local testing.

### Describe your proposed solution

Add an environment variable SKLEARN_DEFAULT_PARALLEL_BACKEND that can be set to the relevant joblib backends (loky/threading/serial).

### Describe alternatives you've considered, if relevant

* An env variable could be added to joblib, and then it would be JOBLIB_DEFAULT_PARALLEL_BACKEND.
* This could be a Python API, but that means e.g. pytest needs to grow more options so it can call the API, and puts (a bit) more burden on users who want to experiment.
* Free-threaded CPython builds could opt-in to thread pools by default. In the long term this may be a better option, in the short term I'd worry about correctness; seems better to change defaults here incrementally.

### Additional context

This breaks out one tiny aspect of #34057, where this was done as part of a flag to `pytest`, and I received feedback disliking such flags. And also this is more generally useful.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [60](https://github.com/godofecht/flow-scikit/issues/60) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [61](https://github.com/godofecht/flow-scikit/issues/61) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
